Adding Treeview Widget

keithknott26 opened the pull request:
Hello Jakob,

Thanks for your hard work on this project, I often think about ways I could use it in my development efforts. Anyway, I've added a Treeview widget, demo, and tests to a fork of your repo under keithknott26/termdash.  I felt I should contribute something hopefully useful to this effort.

Let me know if you'd like me to make any changes, the test coverage is not ideal given I wasn't too familiar with writing them for a UI type application. Mutex locks are in place so there are currently no race conditions as far as I know. Mouse and keyboard integration are included.

Cheers,
Keith Knott

mum4k committed:
Merge pull request #374 from mum4k/release-0-20-0

Releasing Termdash v0.20.0.

keithknott26 committed:
Adding Treeview widget and tests

keithknott26 committed:
Updated testing errors to pass

keithknott26 committed:
Updated treeview and tests to guard against race conditions

keithknott26 committed:
Updated tests and treeview with mutex locking

keithknott26 committed:
Minor fixes for linting errors

mum4k commented:
Hello @keithknott26 and thank you, this looks great! Thank you for adding a new widget and including solid test coverage. It will take me a bit to review this. I will get back to you if I see any points we need to address.

mum4k commented:
I just had some time to play with the demo, the `treeview` widget looks really cool, thank you @keithknott26!

Going to start reviewing it shortly.

mum4k commented:
@keithknott26, while I am reviewing this - do you want to update this PR adding a screenshot of this widget and its capabilities to the main README? It doesn't have to be a moving gif like the other ones, but can be if you want to.

mum4k commented on widgets/treeview/treeview.go:
```diff
@@ -0,0 +1,818 @@
+
+// TreeNode represents a node in the treeview.
+type TreeNode struct {
+	ID            string
```
Would you mind documenting all these attributes? It would help the posterity to maintain this widget.

mum4k commented on widgets/treeview/treeview.go:
```diff
@@ -0,0 +1,818 @@
+
+// SetShowSpinner safely sets the ShowSpinner flag.
+func (node *TreeNode) SetShowSpinner(value bool) {
+	node.mu.Lock()
```
Wherever locking the mutex, consider the following pattern where we defer its release. That guarantees the mutex will be released regardless of how the function exists. The defer method tends to be more3 resilient to future code changes where people could ad additional exists (e.g. an early return) and forget to unlock.

```go
func foo() {
  node.mu.Lock()
  defer node.mu.Unlock()

  // function's logic goes here
}
```

mum4k commented on widgets/treeview/treeview.go:
```diff
@@ -0,0 +1,818 @@
+}
+
+// SetShowSpinner safely sets the ShowSpinner flag.
+func (node *TreeNode) SetShowSpinner(value bool) {
```
It is recommended that the name of the receiver (here `node`) represents the dentity of the type. Consider `tn` or similar instead.

```go
func (tn *TreeNode) ...
```

mum4k commented on widgets/treeview/treeview.go:
```diff
@@ -0,0 +1,818 @@
+
+// GetShowSpinner safely retrieves the ShowSpinner flag.
+func (node *TreeNode) GetShowSpinner() bool {
+	return node.ShowSpinner
```
Since `node.ShowSpinner` is written under a mutex, we should also read it under a mutex for a complete thread safety.

mum4k commented on widgets/treeview/treeview.go:
```diff
@@ -0,0 +1,818 @@
+
+// Treeview represents the treeview widget.
+type Treeview struct {
+	mu                  sync.Mutex
```
Same comment here, if possible, please document all the fields to ease future maintenance.

mum4k commented on widgets/treeview/treeview.go:
```diff
@@ -0,0 +1,818 @@
+			visibleNodes := tv.getVisibleNodesList()
+			tv.mu.Unlock() // Release the Treeview lock before operating on individual nodes
+			for _, node := range visibleNodes {
+				//node.mu.Lock()
```
Did you mean to leave this unlocked? Or should we lock the node here?

mum4k commented on widgets/treeview/treeview.go:
```diff
@@ -0,0 +1,818 @@
+}
+
+// Treeview represents the treeview widget.
+type Treeview struct {
```
Consider changing the name to TreeView to matche the Go's recommended style of mixedCaps.

https://google.github.io/styleguide/go/guide#mixed-caps

mum4k reviewed (changes requested):
Looks great Keith, just a few comments. Most of them optional for your consideration.

keithknott26 committed:
Added missing comments to types,  added more tests, changed node to tv, changed Treeview to TreeView

keithknott26 committed:
Added animated gif and updated README

keithknott26 committed:
Added suimon since it went missing in a previous commit

keithknott26 commented:
Thanks! 
I've incorporated your suggestions and updated the following:
Treeview to TreeView
Added missing comments to types
Changed node to tv
Updates to mutex locking
Added more tests
README.md file was updated with an animated gif

Let me know if you think anything else should be updated.

Cheers,
Keith

mum4k commented:
Looks great, thank you very much for iterating @keithknott26 and for adding the animated gif that shows the main features of this widget. I will review the other widget shortly, so we can release them to the master branch together.
